### PR TITLE
Replacing NotNullAttribute with exceptions

### PR DESCRIPTION
**Pull request by pranavkm**

### CODE REVIEW

**Comment by pranavkm** [comment]
cc @davidfowl \ @Eilon

File: src/Microsoft.AspNet.Routing/Constraints/RegexInlineRouteConstraint.cs [hunk 1]
```diff
@@ -14,9 +14,13 @@ public class RegexInlineRouteConstraint : RegexRouteConstraint
         /// Initializes a new instance of the <see cref="RegexInlineRouteConstraint" /> class.
         /// </summary>
         /// <param name="regexPattern">The regular expression pattern to match.</param>
-        public RegexInlineRouteConstraint([NotNull] string regexPattern)
+        public RegexInlineRouteConstraint(string regexPattern)
             : base(regexPattern)
         {
+            if (regexPattern == null)
```
Comment (dougbu) [inline src/Microsoft.AspNet.Routing/Constraints/RegexInlineRouteConstraint.cs]: Should only need this check in the `base` class.

File: src/Microsoft.AspNet.Routing/Constraints/BoolRouteConstraint.cs [hunk 2]
```diff
@@ -15,12 +14,32 @@ namespace Microsoft.AspNet.Routing.Constraints
...
     {
         /// <inheritdoc />
-        public bool Match([NotNull] HttpContext httpContext,
-                          [NotNull] IRouter route,
-                          [NotNull] string routeKey,
-                          [NotNull] IDictionary<string, object> values,
+        public bool Match(HttpContext httpContext,
+                          IRouter route,
+                          string routeKey,
+                          IDictionary<string, object> values,
```
Comment (dougbu) [inline src/Microsoft.AspNet.Routing/Constraints/BoolRouteConstraint.cs]: pity your compile module doesn't move the first parameter to its own line and fix the indentation of the reset :smiling_imp:

File: src/Microsoft.AspNet.Routing/Template/TemplateRoute.cs [hunk 3]
```diff
@@ -35,9 +34,13 @@ public class TemplateRoute : INamedRouter
                                dataTokens: null,
                                inlineConstraintResolver: inlineConstraintResolver)
         {
+            if (target == null)
```
Comment (dougbu) [inline src/Microsoft.AspNet.Routing/Template/TemplateRoute.cs]: Don't duplicate checks the chained constructor has already done.

File: src/Microsoft.AspNet.Routing/Template/TemplateRoute.cs [hunk 4]
```diff
@@ -45,9 +48,13 @@ public class TemplateRoute : INamedRouter
                              IInlineConstraintResolver inlineConstraintResolver)
             : this(target, null, routeTemplate, defaults, constraints, dataTokens, inlineConstraintResolver)
         {
+            if (target == null)
```
Comment (dougbu) [inline src/Microsoft.AspNet.Routing/Template/TemplateRoute.cs]: As above.

File: src/Microsoft.AspNet.Routing/VirtualPathData.cs [hunk 5]
```diff
@@ -48,10 +55,15 @@ public VirtualPathData([NotNull] IRouter router, string virtualPath)
         /// <param name="virtualPath">The generated URL.</param>
         /// <param name="dataTokens">The collection of custom values.</param>
         public VirtualPathData(
-            [NotNull] IRouter router,
+            IRouter router,
             PathString virtualPath,
             IDictionary<string, object> dataTokens)
         {
+            if (router == null)
```
Comment (dougbu) [inline src/Microsoft.AspNet.Routing/VirtualPathData.cs]: Only need this `null` check in this class.

File: test/Microsoft.AspNet.Routing.Tests/Template/TemplateMatcherTests.cs [hunk 6]
```diff
@@ -1,12 +1,9 @@
 // Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
-#if DNX451
 using System.Collections.Generic;
-using Microsoft.AspNet.Routing.Constraints;
 using Microsoft.Framework.DependencyInjection;
 using Microsoft.Framework.OptionsModel;
-using Moq;
```
Comment (dougbu) [inline test/Microsoft.AspNet.Routing.Tests/Template/TemplateMatcherTests.cs]: :dancers:

**Comment by dougbu** [comment]
:shipit: after removing a few unnecessary checks.

**Comment by Eilon** [comment]
:open_mouth: :shipit:

**Pull request closed without merging**